src/codegen/compile.ml

```diff
@@ -10315,6 +10315,22 @@ and compile_exp_as env ae sr_out e =
   let sr_in, code = compile_exp_with_hint env ae (Some sr_out) e in
   code ^^ StackRep.adjust env sr_in sr_out
 
+and single_case e (cs : Ir.case list) =
+  match cs, e.note.Note.typ with
+  | [{it={pat={it=TagP (l, _);_}; _}; _}], Type.(Variant [{lab; _}]) -> l = lab
+  | _ -> false
+
+and known_tag_pat p = TagP ("", p)
+
+and simplify_cases e (cs : Ir.case list) =
+  match cs, e.note.Note.typ with
+  (* for a 2-cased variant type, the second second comparison can be omitted when the first pattern
+     (with irrefutable subpattern) didn't match *)
+  | [{it={pat={it=TagP (_, ip);_}; _}; _} as c1; {it={pat={it=TagP (_, pat');_} as pat2; exp}; _} as c2], Type.(Variant [_; _])
```

Comment:
This isn't quite enough - I think you need to check the pattern tags are in the variant. See variant-specials-bug.

Reply:
Subtyping is a bitch.

Reply:
The first match's tag shouldn't be assumed to be in the scrutinee's type either! I made two test cases, both work as expected now.